src/codegen/baseline_jit.h

```diff
@@ -0,0 +1,249 @@
+#include "codegen/ast_interpreter.h"
+#include "gc/heap.h"
+
+namespace pyston {
+
+#define ENABLE_BASELINEJIT_ICS 1
+
+class AST_stmt;
+class Box;
+class BoxedDict;
+class BoxedList;
+class BoxedTuple;
+
+class SetGlobalIC;
+class GetGlobalIC;
+class SetAttrIC;
+class GetAttrIC;
+class SetItemIC;
+class GetItemIC;
+class CompareIC;
+class AugBinopIC;
+class UnaryopIC;
+class RuntimeCallIC;
+
+class JitFragment;
+
+// Manages a fixed size memory block which stores JITed code.
+// It can contain a variable number of blocks generated by JitFragment instances.
+// Currently a JitFragment always contains the code of a single CFGBlock*.
+// A JitFragment can get called from the Interpreter by calling 'entry_code' which will jump to the fragment start or
+// it can get executed by a jump from another fragment.
+// At every fragment end we can jump to another fragment, fallback to the Interpreter or exit.
+// This means we are not allowed to assume that a register contains a specific value between JitFragments.
+// This also means that we are allowed to store a Python variable which only lives in the current CFGBLock* inside a
+// register or stack slot but we aren't if it outlives the block - we have to store it in the interpreter instance.
+//
+// To execute a specific CFGBlock one has to call:
+//      CFGBlock* block;
+//      block->entry_code(ast_interpreter_instance, block)
+//
+// Signature of a JitCodeBlock:
+//  std::pair<CFGBlock*, Box*>(*entry_code)(ASTInterpreter* interp, CFGBlock* block)
+//  args:
+//      interp:  instance to the ASTInterpreter
+//      block:   block to execute
+//
+//  return value:
+//      first:   next block to execute in the interpreter
+//                  if 0, we executed a return node and should return second
+//      second:  return value only used if first == 0
+//
+// Basic layout of generated code block is:
+// entry_code:
+//      push   %rbp                 ; setup frame pointer
+//      mov    %rsp,%rbp
+//      sub    $0x108,%rsp          ; setup scratch, 0x108 = scratch_size + 8 (=stack alignment)
+//      push   %rdi                 ; save the pointer to ASTInterpreter instance
+//      jmpq   *0x8(%rsi)           ; jump to block->code
+//                                      possible values: first_JitFragment, second_JitFragment,...
+//
+// first_JitFragment:
+//      ...
+//    ; Side exit (e.g. if <val> <block1> else <block2>
+//      movabs $0x1270014108,%rcx   ; rcx = True
+//      cmp    %rax,%rcx            ; rax == True
+//      jne    end_side_exit
+//      mov    %rax,0x10(%rsp)      ;
+//      movabs $0x215bb60,%rax      ; rax = CFGBlock* to call next (rax is also the 1. return reg)
+//      mov    0x8(%rax),%rsi       ; load CFGBlock->code
+//      test   %rsi,%rsi            ; CFGBlock->code == 0
+//      je     epilog               ; exit to interpreter if code == 0
+//      jmpq   *0x8(%rax)           ; jump to new jit fragment (e.g second_JitFragment)
```

Comment:
also ```emitReturn``` etc should probably just directly emit a ```leave``` and ```ret``` and not jump to the epilog because the jump takes more bytes and is an additional branch.